### DISCUSSION TITLE
Get values of different nodes in subscription

### DISCUSSION THREAD

**Discussion by Jack9253** [FreeOpcUa/python-opcua#1196] [Q&A]
Hi everybody,

I,m new to Python so please be kind.
I got an OPC Sever with several nodes. I want to read them by using subscription as it seems to be the fastest way.
Values are changing in invervals of 5-20 ms. I've set up and subscription handler and everything is working just fine.

But I'm wondering, how to check, which node triggered the data_change function. I want to append the values of the different nodes to individual variables to save them afterwards. 

How can i check which nodes triggers the data_change? Do i really have to parse all the Node String? All my nodes are around 20-30 signs and not of equal length. So would be quite complicated. 

Thanks a lot
Jack

`from opcua import Client
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import time

valuesN1 = []
valuesN2  = []
valuesN3  = []

class SubHandler(object):

    def datachange_notification(self, node, val, data):

      __ here i need to check which node triggered the function, to write the value to the correct variable __

      valuesN1.append(val) # IF N1
      valuesN2.append(val) # IF N2
      valuesN3.append(val) # IF N3

      print(val)
       
client = Client("my_URL_XYZ")
client.connect()

N1  =   client.get_node("ns=12;s=XXX_YYY_ZZZ")
N2  =   client.get_node("ns=12;s=XX_YY_ZZ")
N3  =   client.get_node("ns=12;s=X_Y_Z")

data = [N1,N2,N3]

handler = SubHandler()
sub = client.create_subscription(10,handler)
handle = sub.subscribe_data_change(data)`

**Comment by AndreasHeine** [comment]
first of all publishing intervals lower then 50ms are kind of unrealistic in python especialy if you plan to hace many of them!
second yes you need to sort them!

**Reply by Jack9253** [reply]
Thank you very much for your reply.
I tried to sort the nodes in the data change function like so 

`class SubHandler(object):

    def datachange_notification(self, node, val, data):
      print('test')
      
      if node == Node_N1:
           print('test_N1')      
           N1_values.append(val)

      if node == Node_N2:
           print('test_N2')      
           N2_values.append(val)`

In Node_N1 and Node_N2 the identifier strings of the nodes are stored. I printed them with print(node) and checked if they are equal.
If a data change occurs only '__test__' ist printed, so the comparison of the strings seems not to work correctly. Is this the wrong way?

Is there a better way to read values at fixed intervals? I dont need to catch every data change, but I want to get values of a fixed number of nodes at certain intervals. For example every 50 ms the values of 10 nodes should be read and stored. 

Thanks for your support!

**Reply by AndreasHeine** [reply]
node is a object from type NodeClass https://github.com/FreeOpcUa/python-opcua/blob/1cca33dd5fb70858b5e811b08356ca97520bcbaf/opcua/common/node.py#L31

i would recommend you comparing the nodeids or browsenames and sort them by so (storing them in a dictionary with nodeid as key is also a good way to start)

**Reply by AndreasHeine** [reply]
> Is there a better way to read values at fixed intervals? I dont need to catch every data change, but I want to get values of a fixed number of nodes at certain intervals. For example every 50 ms the values of 10 nodes should be read and stored.

a subscription is the most efficient way! but it is not read in fixed intervals!!! 50ms is only the interval of picking up the notifications if any datachange happened since last publish.

if you want to read values cyclic ("polling") you need to read in a while loop and a sleeptime (but this is not efficient if you have many nodes)
there are two mainways to do so:
NodeClass has a get_value method (reads the value of a single node)
https://github.com/FreeOpcUa/python-opcua/blob/1cca33dd5fb70858b5e811b08356ca97520bcbaf/opcua/common/node.py#L148

ClientClass has a get_values method which reads many nodes in one request <- you should look at this if you dont want to subscribe them!
https://github.com/FreeOpcUa/python-opcua/blob/1cca33dd5fb70858b5e811b08356ca97520bcbaf/opcua/client/client.py#L658
